### PR TITLE
Data models and custom methods

### PR DESCRIPTION
**Pull request by KatIsCoding**
In this update, the data models and their custom methods have been implemented. Some of the custom methods implemented are:
- Method that updates the comment counter for a post
- Method that updates the likes counter for a post
- Method that updates the post counter for a given user.
- Method that returns the most recent comments in a post
- A method that returns the 3 most recent posts for a given user.

### CODE REVIEW

File: app/models/comment.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class Comment < ActiveRecord::Base
+  # Method that updates the comment counter for a post
```
Comment (sinansevgi) [inline app/models/comment.rb:4]: - [ ] The associations between the comment model and `user and post` models should be set up correctly.

```ruby
belongs_to :author, class_name: 'User'
belongs_to :post
```

File: app/models/like.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Like < ActiveRecord::Base
+  # Method that updates the likes counter for a post
```
Comment (sinansevgi) [inline app/models/like.rb:3]: - [ ] The associations between the like model and `user and post` models should be set up correctly.

File: app/models/post.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class Post < ActiveRecord::Base
+  # Method that updates the post counter for a given user.
```
Comment (sinansevgi) [inline app/models/post.rb:5]: - [ ] The associations between the post model and other models should be set up correctly.

```ruby
belongs_to :author, class_name: 'User'
has_many :comments
has_many :likes
```

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class User < ActiveRecord::Base
+  # A method that returns the 3 most recent posts for a given user.
```
Comment (sinansevgi) [inline app/models/user.rb]: - [ ] The associations between the user model and other models should be set up correctly.

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  # The author_id is passed in as a parameter.
+  #
+  # @param author_id [Integer]
+  # @return [Array]
+  #
+  def self.recent_posts(user:)
+    Post.where(author_id: user.id).order(created_at: :desc).limit(3)
+  end
```
Comment (sinansevgi) [inline app/models/user.rb]: - [ ] You should make queries through the association. 

```ruby
def recent_post
  posts.order('created_at Desc').limit(3)
end
```

File: app/models/post.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  def self.update_counter(user:)
+    user.update(postscounter: Post.where(author_id: user.id).count)
+  end
+
+  # Method that returns the most recent comments in a post
+
+  def self.recent_comments(post:)
+    Comment.where(post_id: post.id).order(created_at: :desc).limit(5)
+  end
+end
```
Comment (sinansevgi) [inline app/models/post.rb]: - [ ] You should make queries through the association. If you correctly configured your associations, you should be able to access comments that belong to the post without making extra SQL queries.

File: app/models/post.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+class Post < ActiveRecord::Base
+  # Method that updates the post counter for a given user.
+
+  def self.update_counter(user:)
+    user.update(postscounter: Post.where(author_id: user.id).count)
+  end
```
Comment (sinansevgi) [inline app/models/post.rb]: - [ ] You should make queries through the association.

File: app/models/post.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+class Post < ActiveRecord::Base
+  # Method that updates the post counter for a given user.
+
+  def self.update_counter(user:)
```
Comment (sinansevgi) [inline app/models/post.rb]: [Optional]
- [ ] You can user [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) to update the counter automatically after a new post created

```ruby
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
end
```

File: app/models/like.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+
+  def self.update_counter(post:)
+    post.update(likescounter: Like.where(post_id: post.id).count)
+  end
```
Comment (sinansevgi) [inline app/models/like.rb]: - [ ] You should make queries through the association.

File: app/models/comment.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+
+  def self.update_counter(post:)
+    post.update(comments_count: Comment.where(post_id: post.id).count)
+  end
```
Comment (sinansevgi) [inline app/models/comment.rb]: - [ ] You should make queries through the association.
```ruby
def update_comments_counter
  post.increment!(:comments_counter)
end
```

**Review by sinansevgi — CHANGES_REQUESTED** [review]
Hi @KatIsCoding , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
- [ ] You should create a method to set `current_user` in your `ApplicationController` as requested [in the project specifications](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project.md#:~:text=Additionally%2C%20you%20will,the%20upcoming%20projects.).

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Happy Hacking! :computer: 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.
Sinan Sevgi

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi @KatIsCoding, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**